sycl/source/detail/graph_impl.cpp

```diff
@@ -701,16 +658,23 @@ void graph_impl::makeEdge(std::shared_ptr<node_impl> Src,
   // We need to add the edges first before checking for cycles
   Src->registerSuccessor(Dest);
 
+  // Dest is no longer a Root node, so we need to remove it from MRoots.
+  if (Dest->MPredecessors.size() == 1) {
+    MRoots.erase(Dest);
+  }
+
   // We can skip cycle checks if either Dest has no successors (cycle not
   // possible) or cycle checks have been disabled with the no_cycle_check
   // property;
   if (Dest->MSuccessors.empty() || !MSkipCycleChecks) {
     bool CycleFound = checkForCycles();
 
     if (CycleFound) {
-      // Remove the added successor and predecessor
+      // Remove the added successor and predecessor.
       Src->MSuccessors.pop_back();
       Dest->MPredecessors.pop_back();
+      // Add the Dest back into MRoots.
+      MRoots.insert(Dest);
```

Comment:
If `Dest` always a root node in this control flow? in `if (Dest->MSuccessors.empty() || !MSkipCycleChecks) {` we could get to this point if the second or condition is true but not the first.

Reply:
I think this was a bug. Thanks for noticing. I added more checks to make sure that Dest is actually a root node and that the edge gets added successfully.